File: C++/map.cpp
#include <iostream>
#include <conio.h>
#include <string>

#define KEY_UP 72
#define KEY_DOWN 80
#define KEY_LEFT 75
#define KEY_RIGHT 77

#if defined(WIN32) || defined(_WIN32) || defined(__WIN32) && !defined(__CYGWIN__)
    #define CLEAR system("cls")
#else
    #define CLEAR system("clear")
#endif

void map(int size, int row, int column)
{
    std::cout << "MAP " << std::string(((size * 2) - 5), ' ') << row << 'x' << column << std::endl;

    for (int axis = 1; axis <= size; axis++) {
        if (axis == 1 || axis == size)
            std::cout << '+' << std::string((size * 2), '-') << '+' << std::endl;
        else if (axis == (row + 1))
            std::cout << '|' << std::string(((column) * 2), ' ') << "\e[31mp\e[0m" << std::string((((size - column) * 2) - 1), ' ') << '|' << std::endl;
        else
            std::cout << '|' << std::string((size * 2), ' ') << '|' << std::endl;
    }
}

int main()
{
    int key = 0,
        size = 15,
        x, y;

    y = x = (size / 2);

    CLEAR;
    map(size, y, x);

    while(1) {
        key = 0;

        switch ((key = getch())) {
        case KEY_UP:
            if (y != 1) y--;
            break;
        case KEY_DOWN:
            if (y != (size - 2)) y++;
            break;
        case KEY_LEFT:
            if (x != 0 ) x--;
            break;
        case KEY_RIGHT:
            if (x != (size - 1)) x++;
            break;
        case 'q':
            exit(1);
        default:
            break;
        }

        CLEAR;
        map(size, y, x);
    }

    return 0;
}

File: Ruby/map.rb
require 'io/console'

module OS
  class << self
    def windows?
      return !!(/cygwin|mswin|mingw|bccwin|wince|emx/ =~ RUBY_PLATFORM)
    end

    def mac?
      return !!(/darwin/ =~ RUBY_PLATFORM)
    end

    def unix?
      return !windows?
    end

    def linux?
      return unix? if !mac?
    end
  end
end

def clear
  OS.windows? ? system('cls') : system('clear')
end

clear()

@fruit = [4,1]

def map size, row, column
  print "MAP %#{size-2}dx%d\n" % [column, row]

  axis = 1

  while axis != 10 do
    # fruit = ' ' * (rand(1..size).freeze)
    if axis == 1 || axis == 9
      print "+ #{'-' * size} +\n"
    elsif axis == row
      print "#{"\e[47m|\e[0m"} #{' ' * (column - 1) + "\e[31m♦\e[0m" + ' ' * (size - column)} |\n"
    else
      print "| #{' ' * (size)} |\n"
    end
    axis += 1
  end
end

def read_char
  STDIN.echo = false
  STDIN.raw!

  input = STDIN.getc.chr
  if input == "\e" then
    input << STDIN.read_nonblock(3) rescue nil
    input << STDIN.read_nonblock(2) rescue nil
  end
ensure
  STDIN.cooked!

  return input
end

pos      = 1
map_size = 30
y_axis   = 5
x_axis   = map_size / 2

map(map_size, y_axis, x_axis)

while (y_axis != pos)
  char = read_char

  case char
  when "\u0003"
    exit 0
  when /^.$/
    if char == ?s
      y_axis += 1 unless y_axis >= 8
    elsif char == ?w
      y_axis -= 1 unless y_axis <= 2
    elsif char == ?d
      x_axis += 1 unless x_axis >= map_size
    elsif char == ?a
      x_axis -= 1 unless x_axis <= 1
    end

    clear() if y_axis != (pos - 1)
  end

  map(map_size, y_axis, x_axis)
  $stdout.flush
end

